### PR TITLE
Add Kinesis and DynamoDB event sources

### LINKED ISSUE #2365: New kinesis syntax not working
# This is a (Bug Report / Feature Proposal)
## Description

For bug reports:
- What went wrong?
  When I tried to convert my kinesis event mappings to the new syntax, like this:

`````` yaml

  extract:
    handler: extract/index.handler
    events:
      - stream:
          arn:
            Fn::GetAtt:
              - BookmarksStream
              - Arn

I got the following error "EventSourceArn.split is not a function"

Looking at the source code, I get the impression that this feature simply has not been implemented yet. However, it is already referenced in the docs here:

https://serverless.com/framework/docs/providers/aws/events/streams/

* What did you expect should have happened?

The cloudformation GetAtt should have resolved into a valid arn

* What was the config you used?

```yaml

  extract:
    handler: extract/index.handler
    events:
      - stream:
          arn:
            Fn::GetAtt:
              - BookmarksStream
              - Arn

* What stacktrace or error message from your provider did you see?

For feature proposals:
* What is the use case that should be solved. The more detail you describe this in the easier it is to understand for us.
* If there is additional config how would it look

Similar or dependent issues:
* #12345

## Additional Data

* ***Serverless Framework Version you're using***:
1.0.2 

* ***Operating System***:

linux (kubuntu)
* ***Stack Trace***:

TypeError: EventSourceArn.split is not a function
    at functionObj.events.forEach.event (/home/yeehaa/.nvm/versions/node/v6.8.0/lib/node_modules/serverless/lib/plugins/aws/deploy/compile/events/stream/index.js:82:47)
    at Array.forEach (native)
    at serverless.service.getAllFunctions.forEach (/home/yeehaa/.nvm/versions/node/v6.8.0/lib/node_modules/serverless/lib/plugins/aws/deploy/compile/events/stream/index.js:20:28)
    at Array.forEach (native)
    at AwsCompileStreamEvents.compileStreamEvents (/home/yeehaa/.nvm/versions/node/v6.8.0/lib/node_modules/serverless/lib/plugins/aws/deploy/compile/events/stream/index.js:16:47)
    at BbPromise.reduce (/home/yeehaa/.nvm/versions/node/v6.8.0/lib/node_modules/serverless/lib/classes/PluginManager.js:151:50)
    at tryCatcher (/home/yeehaa/.nvm/versions/node/v6.8.0/lib/node_modules/serverless/node_modules/bluebird/js/release/util.js:16:23)
    at Object.gotValue (/home/yeehaa/.nvm/versions/node/v6.8.0/lib/node_modules/serverless/node_modules/bluebird/js/release/reduce.js:157:18)
    at Object.gotAccum (/home/yeehaa/.nvm/versions/node/v6.8.0/lib/node_modules/serverless/node_modules/bluebird/js/release/reduce.js:144:25)
    at Object.tryCatcher (/home/yeehaa/.nvm/versions/node/v6.8.0/lib/node_modules/serverless/node_modules/bluebird/js/release/util.js:16:23)
    at Promise._settlePromiseFromHandler (/home/yeehaa/.nvm/versions/node/v6.8.0/lib/node_modules/serverless/node_modules/bluebird/js/release/promise.js:510:31)
    at Promise._settlePromise (/home/yeehaa/.nvm/versions/node/v6.8.0/lib/node_modules/serverless/node_modules/bluebird/js/release/promise.js:567:18)
    at Promise._settlePromise0 (/home/yeehaa/.nvm/versions/node/v6.8.0/lib/node_modules/serverless/node_modules/bluebird/js/release/promise.js:612:10)
    at Promise._settlePromises (/home/yeehaa/.nvm/versions/node/v6.8.0/lib/node_modules/serverless/node_modules/bluebird/js/release/promise.js:691:18)
    at Async._drainQueue (/home/yeehaa/.nvm/versions/node/v6.8.0/lib/node_modules/serverless/node_modules/bluebird/js/release/async.js:138:16)
    at Async._drainQueues (/home/yeehaa/.nvm/versions/node/v6.8.0/lib/node_modules/serverless/node_modules/bluebird/js/release/async.js:148:10)

* ***Provider Error messages***:

EventSourceArn.split is not a function
``````

### PR DESCRIPTION
**Pull request by eahefnawy**
## What did you implement:

Closes #2365

Added new `-kinesis` and `-dynamodb` events while keeping `-stream` for backward compatibility. This makes it easy for us to detect the type without parsing the ARN. Which means we can take the ARN as it is and pass it to CF which fixes issue #2365

Other than that, the logical ID of the stream is now using an iterator for each type instead of the stream name to drop the dependency on the ARN parsing.

Here's the Pr for the `stream` event: https://github.com/serverless/serverless/pull/2250
## How can we verify it:

To verify, run `serverless deploy --noDeploy` on the following configs and check the update template.
### Verifying old stream syntax works

create a dynamodb table & stream and use the following config:

``` yml
service: serverless-stream-type

provider:
  name: aws
  runtime: nodejs4.3

functions:
  hello:
    handler: handler.hello
    events:
      - stream: arn:aws:dynamodb:us-east-1:123456789:table/stream-type/stream/2016-10-20T15:04:38.834
```
### Verifying dynamodb stream syntax works

create a dynamodb table & stream and use the following config:

``` yml
service: serverless-stream-type

provider:
  name: aws
  runtime: nodejs4.3

functions:
  hello:
    handler: handler.hello
    events:
      - dynamodb: arn:aws:dynamodb:us-east-1:123456789:table/stream-type/stream/2016-10-20T15:04:38.834
```
### Verifying kinesis stream syntax works

create a kinesis stream and use the following config:

``` yml
service: serverless-stream-type

provider:
  name: aws
  runtime: nodejs4.3

functions:
  hello:
    handler: handler.hello
    events:
      - kinesis: arn:aws:kinesis:us-east-1:123456789:table/stream-type/stream/2016-10-20T15:04:38.834
```
## Todos:
- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config/commands/resources
- [x] Change ready for review message below

**_Is this ready for review?:**_ YES

### CODE REVIEW

**Comment by pmuens** [comment]
👍  good job. Will review thoroughly tomorrow morning!

Can you add a `console.log` statement that the user should update to the `kinesis` or `dynamodb` event source if he still uses `stream` so that we can deprecate it soon?

**Comment by pmuens** [comment]
@eahefnawy just a friendly reminder to think about this document as well when you update the docs: https://github.com/serverless/serverless/blob/master/docs/02-providers/aws/04-resource-names-reference.md

**Comment by pmuens** [comment]
Just ran my first test on this one with the following `serverless.yml`:

```yml
service: aws-nodejs

provider:
  name: aws
  runtime: nodejs4.3

functions:
  hello:
    handler: handler.hello
    events:
      #- stream: arn:aws:dynamodb:us-east-1:XXXX:table/stream-event/stream/2016-10-21T03:46:11.547
      #- stream: arn:aws:kinesis:us-east-1:XXXX:stream/stream-event
      - stream:
          arn: arn:aws:dynamodb:us-east-1:XXXX:table/stream-event/stream/2016-10-21T03:46:11.547
          batchSize: 20
          startingPosition: TRIM_HORIZON
          enabled: true
      - stream:
          arn: arn:aws:kinesis:us-east-1:XXXX:stream/stream-event
          batchSize: 15
          startingPosition: LATEST
          enabled: false
```

At first I used the simple string representation which worked great. After that I decided to test the object representation and deployed again with the same stream. It broke there as the mapping is already defined.

Here's the error message:

```
 An error occurred while provisioning your stack: HelloEventSourceMappingKinesis1
     - The event source arn (arn:aws:kinesis:us-east-1:XXXX:stream/stream-event)
      and function (aws-nodejs-dev-hello) provided mapping
     already exists. Please update or delete the existing
     mapping with UUID d8b83ad5-7d06-4683-955d-97a58de1544d.
```

Maybe we should switch from a number for the streams to another unique item for the ResourceId. This might also break if I change the order of the stream definition.

/cc @flomotlik

**Comment by eahefnawy** [comment]
@pmuens did you use the same stream mapping twice for the same function? I don't think you can do that ... or you mean something else?

**Comment by pmuens** [comment]
I switched from string to object syntax to configure it in detail.

I commented out the string syntax and wrote it in object syntax like it's shown above.

**Comment by flomotlik** [comment]
@eahefnawy and this also works with GetAtt function?

**Comment by eahefnawy** [comment]
@flomotlik yes it does, but of course it only works with `kinesis` and `dynamodb` events, but not the deprecated `stream` event for the reasons discussed earlier.

@pmuens FYI I was able to reproduce the issue with **master** branch as well. So it has nothing to do with the iterator or this PR in general. For some reason, CF is trying to recreate the event source mapping rather than updating, which fails because it already exists. So I'm not sure how to proceed on this front, but imo it's irrelevant to this PR.

**Comment by pmuens** [comment]
It seems like we're missing a `DependsOn` or something as the first deployment always fails on my machine with the following error message:

```
Serverless Error ---------------------------------------
 
     An error occurred while provisioning your stack: HelloEventSourceMappingKinesis1
     - Cannot access stream arn:aws:kinesis:us-east-1:XXXX:stream/stream-event.
     Please ensure the role can perform the GetRecords, GetShardIterator,
     DescribeStream, and ListStreams Actions on your stream
     in IAM..
```

However the second deployment succeeds. It seems like the permissions are not in place after the first deployment.

**Comment by yeehaa123** [comment]
@pmuens I ran into the same issue yesterday.

**Comment by eahefnawy** [comment]
@pmuens @yeehaa123 hmmm I can't reproduce the issue. Deployment succeeds on first trial for me. [There's already a `DependsOn` property for the policy on the template](https://github.com/serverless/serverless/pull/2488/files#diff-ec6453951d23addb4431d1b1a3556796R84) so this should work. My best guess it might be an CF issue?

**Comment by pmuens** [comment]
Hmm. That's interesting. We faced the problem multiple times yesterday at the workshop.

Another way to reproduce it is to clone this repo: https://github.com/ServerlessInc/serverlessconf-workshop

And then do a `serverless deploy`. This will create a Kinesis stream through custom resources and then tries to create the event source mapping (through the `stream` event). That should fail and cause the error above.

**Comment by eahefnawy** [comment]
@pmuens ahaaa you mean this happens when we create a stream using custom resources! so this `DependsOn` should be on the stream not the policy resource.

Hmmm, adding a `DependOn` the stream created with custom resources means that we'll need to detect and parse `GetAttr` to get the logical ID of the stream, which kinda defeats the purpose of this whole implementation.

@flomotlik what you think?

**Comment by eahefnawy** [comment]
@pmuens @flomotlik FYI this might be related to this issue:
https://github.com/serverless/serverless/issues/1981

**Comment by eahefnawy** [comment]
@flo @pmuens FYI I’ve tested the getAttr issue on the workshop repo couple of times and it’s working fine once you change to this:

```yml
 persistVisit:
    handler: persistVisit/index.persistVisit
    events:
      - kinesis:
          arn:
            Fn::GetAtt:
              - WorkshopStream
              - Arn
          batchSize: 1
```
You guys have any other updates? otherwise it's G2M 👍

File: lib/plugins/aws/deploy/compile/events/stream/index.js [hunk 1]
```diff
@@ -15,41 +15,60 @@ class AwsCompileStreamEvents {
...
+          if (event.kinesis) streamType = 'kinesis';
           if (event.stream) {
+            streamType = 'stream';
+            if (!deprecationMessagePrinted) {
+              const warningMessage = [
+                'Deprecation Notice: "stream" event source is deprecated and is now',
+                ' replaced with "dynamodb" or "kinesis" event sources.',
+                ' Please check the docs for more info.',
+              ].join('');
+              console.log(warningMessage); // eslint-disable-line no-console
```
Comment (pmuens) [inline lib/plugins/aws/deploy/compile/events/stream/index.js]: This could re-use the `log` method in `CLI.js`.

**Comment by pmuens** [comment]
Great. Thanks for the updates. Would like to test it thoroughly today before we merge this into master.

**Comment by eahefnawy** [comment]
@pmuens cool ... we need to merge invoke local first and rebase because they include the eslint fixes ... dont wanna fix it twice and conflict

**Comment by pmuens** [comment]
Just tested it again today.
Unfortunately I get this error when I deploy a service from scratch.

```
An error occurred while provisioning your stack: HelloEventSourceMappingDynamodb1
     - Cannot access stream arn:aws:dynamodb:us-east-1:xxxx:table/stream-event/stream/2016-10-21T03:46:11.547.
     Please ensure the role can perform the GetRecords, GetShardIterator,
     DescribeStream, and ListStreams Actions on your stream
     in IAM..
```

The corresponding `serverless.yml` file:

```yml
service: stream-test

provider:
  name: aws
  runtime: nodejs4.3

functions:
  hello:
    handler: handler.hello
    events:
      - stream: arn:aws:dynamodb:us-east-1:xxxx:table/stream-event/stream/2016-10-21T03:46:11.547
      - kinesis: arn:aws:kinesis:us-east-1:xxxx:stream/stream-event
```

It seems like the permissions are not in place for the first deployment.

**note**: it doesn't matter if I use `stream`, `dynamodb` or `kinesis`. It affects all event definitions.

The deployment succeeds after another `deploy`.

**Comment by yeehaa123** [comment]
Is there any news on this one? Anything I can do to move it forward?

**Comment by eahefnawy** [comment]
@yeehaa123 this PR is currently blocked by the bug reported by @pmuens above. I've easily reproduced it. We think it might be an AWS bug, but if you could test out this PR and and hopefully kill the bug, we'd merge right away!! ☺️

**Comment by andrewcurioso** [comment]
@eahefnawy @yeehaa123 any word on this? Is anyone actively working on fixing it?

**Comment by pmuens** [comment]
@andrewcurioso thanks for asking!
This PR is currently blocked because of a kind of race condition during resource creation. We think it's a problem on AWS end.

@eahefnawy Are there any updates from AWS about this?

**Comment by pmuens** [comment]
Maybe a `DependsOn` will help us here. Moving this to the next milestone as I think we need to have a fix for this in place soon.

**Comment by nicka** [comment]
+1 on this PR. BTW got the same issue with IAM policy race conditions. We really need to start testing with putting IAM policies inline! Also my tests with the WaitConditions are still not perfect. Only way to fix it would be with a custom CloudFormation resource backed by a Lambda function, or stress it towards AWS.

**Comment by fleeco** [comment]
+1 on this PR as well.  Started fixing the Fn::GetAttr issue myself then added support for the streams.  Basically have the same implementation in place as what's here - except I added a "Type" property to the stream object.

I like Kinesis and DynamoDB better anyhoo.  Any update on this getting pushed though?  Happy to help!

**Comment by pmuens** [comment]
@fleeco great. 👍 
Yes, we really need to find a good solution here. I think @nicka has something in the pipeline (e.g. putting IAM policies inline) like he has mentioned above.

Any help here is highly appreciated since this kind of race condition problem is pretty hard to get right 💯 🎉 

What do you guys think should be the next step here so that we can move this forward?

**Comment by nicka** [comment]
@pmuens I really think we should go with inline policies, I'll try to come up with something next week! Other than that, I don't see any reason why this PR could not be continued on 👍  Currently we're doing CloudFormation property overwrites to get this working:

```yml
# serverless.yml
publishEvents:
    handler: handler.publishEvents
    events:
      # TODO: Fix streams arns when serverless PR #2488 is merged
      # https://github.com/serverless/serverless/pull/2488
      - stream:
          arn: arn:aws:dynamodb:*:*:table/foo/stream/*
          batchSize: 100
          startingPosition: LATEST
          enabled: true
```

```yml
# cloudformation.yml
  # TODO: Remove when serverless PR #2488 is merged
  # https://github.com/serverless/serverless/pull/2488
  # Associate Lambda functions with streams
  PublishEventsEventSourceMappingDynamodbFoo:
     Properties:
       EventSourceArn:
         Fn::GetAtt:
          - BandsTable
          - StreamArn
```

**Comment by pmuens** [comment]
Great to hear that @nicka! 👍 
And thanks again for helping out here!
Thanks for your example code which resolves this issue!

**Comment by ldadams** [comment]
Looks like this is really close.  @pmuens any thoughts when we might see it in a release?  Thanks!

**Comment by nicka** [comment]
@ldadams @pmuens I guess this PR was somewhat halted because of the IAM timing issues which should be fixed with #2983. Once merged, I'd really would like to see this PR merged 😊

**Comment by pmuens** [comment]
Yes, @ldadams @nicka PR will very likely make this PR mergeable 👍 🎉

**Comment by celalo** [comment]
Do it means this is not going to make it to v1.5, then 😢? #2983 is due for v2.0

**Comment by pmuens** [comment]
Closing this one since #3111 and #2983 were merged a while back and resolved the issues describes here.

**Pull request closed without merging**